financial-templates-lib/test/UniswapPriceFeed.js

```diff
@@ -36,7 +39,68 @@ contract("UniswapPriceFeed.js", function(accounts) {
     await uniswapPriceFeed._update();
 
     assert.equal(uniswapPriceFeed.getCurrentPrice(), null);
+    assert.equal(uniswapPriceFeed.getCurrentTwap(), null);
   });
 
+  // Basic test to verify TWAP (non simulated time).
+  it("Basic 2-price TWAP", async function() {
+    // Update the prices with a small amount of time between.
+    const result1 = await uniswapMock.setPrice(toWei("1"), toWei("1"));
+    await delay(1000);
+    const result2 = await uniswapMock.setPrice(toWei("2"), toWei("1"));
+
+    const getBlockTime = async result => {
+      return (await web3.eth.getBlock(result.receipt.blockNumber)).timestamp;
+    };
+
+    // Grab the exact blocktimes.
+    const time1 = await getBlockTime(result1);
+    const time2 = await getBlockTime(result2);
+    mockTime = time2 + 1;
+
+    // Allow the library to compute the TWAP.
+    await uniswapPriceFeed._update();
+
+
+    const totalTime = mockTime - time1;
+    const weightedPrice1 = toBN(toWei("1")).muln(time2 - time1);
+    const weightedPrice2 = toBN(toWei("0.5")); // 0.5 * 1 second since the mockTime is one second past time2.
+
+    // Compare the TWAPs.
+    assert.equal(uniswapPriceFeed.getCurrentTwap().toString(), weightedPrice1.add(weightedPrice2).divn(totalTime).toString());
+  });
+
+  it("All events before window", async function() {
+    await uniswapMock.setPrice(toWei("1"), toWei("1"));
+    await uniswapMock.setPrice(toWei("2"), toWei("1"));
+    await uniswapMock.setPrice(toWei("4"), toWei("1"));
+
+    // Set the mock time to very far in the future.
+    mockTime = MAX_SAFE_JS_INT;
+
+    await uniswapPriceFeed._update();
+
+    // Expect that the TWAP is just the most recent price, since that was the price throughout the current window.
+    assert.equal(uniswapPriceFeed.getCurrentTwap().toString(), toWei("0.25"));
```

Comment:
One clarifying question: because the `mockTime` is so far in the future the when this is calculated the current TWAP value is simply the last price added to uniswap as this is the only price input over the current window.

Reply:
Exactly. If the last event was before the window (1 hour) started, then the TWAP is just that price weighted by the entire time in the window, since that was the price corresponding to every time point. Does that make sense?